File: medrec/textract.py
import boto3
import io
import sys

word_filter = lambda block: 'BlockType' in block and 'Text' in block and block['BlockType'] == 'WORD'
linebreak = "\n"

def main():
    text = textract()
    result = comprehend(text)

    file = open("py_results.txt", "w")
    file.write(text + linebreak + linebreak)
    file.write(result)
    file.close()

def textract():
    all_text = ''

    s3_conn = boto3.resource('s3')
    s3_obj = s3_conn.Object('just-this-bucket-you-know', 'test_file.png')
    s3_resp = s3_obj.get()

    stream = io.BytesIO(s3_resp['Body'].read())
    image = stream.getvalue()

    client = boto3.client('textract')
    textract_resp = client.analyze_document(Document={'Bytes': image}, FeatureTypes=['TABLES', 'FORMS'])

    blocks = textract_resp['Blocks']
    for block in filter(word_filter, blocks):
        all_text += ' ' + str(block['Text'])

    return all_text

def comprehend(text):
    results = ''

    client = boto3.client('comprehendmedical')
    entity_resp = client.detect_entities(Text = text)
    icd10_resp = client.infer_icd10_cm(Text = text)
    rxnorm_resp = client.infer_rx_norm(Text = text)

    for entity in entity_resp['Entities']:
        results += print_entity(entity)
        if 'Attributes' in entity:
            for attribute in entity['Attributes']:
                results += "    Text:   " + attribute['Text'] + linebreak
                results += "      Type: " + attribute['Type'] + linebreak

    for entity in icd10_resp['Entities']:
        results += print_entity(entity)
        results += print_concept(entity, 'ICD10CMConcepts')

    for entity in rxnorm_resp['Entities']:
        results += print_entity(entity)
        results += print_concept(entity, 'RxNormConcepts')

    return results

def print_entity(entity):
    return "Category:   " + entity['Category'] + linebreak + "  Text:     " + entity['Text'] + linebreak

def print_concept(entity, concept):
    text = ''
    if concept in entity:
        for code in entity[concept]:
            text += "    Code:   " + code['Code'] + linebreak
            text += "      Desc: " + code['Description'] + linebreak
    return text

if __name__ == "__main__":
    main()

File: medrec/textract.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"

	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/comprehendmedical"
	"github.com/aws/aws-sdk-go/service/textract"
)

var linebreak = "\r\n"

func main() {
	defer func() {
		if oof := recover(); oof != nil {
			fmt.Printf("%+v", oof)
		}
	}()

	session := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	},
	))

	text, err := extractText(session)
	errorCheck(err)
	results, err := compMedical(session, text)
	errorCheck(err)

	file, err := os.Create("go_results.txt")
	defer file.Close()
	errorCheck(err)
	file.WriteString(text + linebreak + linebreak)
	file.WriteString(results)
}

func extractText(session *session.Session) (text string, err error) {
	// var s3obj textract.S3Object
	var document textract.Document
	var analyze textract.AnalyzeDocumentInput
	// s3obj.SetBucket("just-this-bucket-you-know")
	// s3obj.SetName("test_file.png")
	// document.SetS3Object(&s3obj)

	file, _ := os.Open("test_file.png")
	reader := bufio.NewReader(file)
	image, _ := ioutil.ReadAll(reader)

	document.SetBytes(image)
	analyze.SetDocument(&document)
	forms := textract.FeatureTypeForms
	tables := textract.FeatureTypeTables
	analyze.SetFeatureTypes([]*string{&forms, &tables})

	service := textract.New(session)
	output, err := service.AnalyzeDocument(&analyze)
	if err != nil {
		return
	}

	for i := range output.Blocks {
		block := output.Blocks[i]
		if *block.BlockType == textract.BlockTypeWord {
			text += " " + *block.Text
		}
	}

	return
}

func compMedical(session *session.Session, text string) (results string, err error) {
	service := comprehendmedical.New(session)
	var dei comprehendmedical.DetectEntitiesV2Input
	var icd10i comprehendmedical.InferICD10CMInput
	var rxni comprehendmedical.InferRxNormInput
	dei.SetText(text)
	icd10i.SetText(text)
	rxni.SetText(text)

	entities, err := service.DetectEntitiesV2(&dei)
	if err != nil {
		return
	}

	for i := range entities.Entities {
		entity := entities.Entities[i]
		results += printEntity(*entity.Category, *entity.Text)
		for j := range entity.Attributes {
			attribute := entity.Attributes[j]
			results += "    Text:   " + *attribute.Text + linebreak
			results += "      Type: " + *attribute.Type + linebreak
		}
	}

	icd10, err := service.InferICD10CM(&icd10i)
	if err != nil {
		return
	}

	for i := range icd10.Entities {
		entity := icd10.Entities[i]
		results += printEntity(*entity.Category, *entity.Text)
		for j := range entity.ICD10CMConcepts {
			concept := entity.ICD10CMConcepts[j]
			results += printConcept(*concept.Code, *concept.Description)
		}
	}

	rxnorm, err := service.InferRxNorm(&rxni)
	if err != nil {
		return
	}

	for i := range rxnorm.Entities {
		entity := rxnorm.Entities[i]
		results += printEntity(*entity.Category, *entity.Text)
		for j := range entity.RxNormConcepts {
			concept := entity.RxNormConcepts[j]
			results += printConcept(*concept.Code, *concept.Description)
		}
	}

	return
}

func printEntity(category string, text string) (result string) {
	return "Category:   " + category + linebreak + "  Text:     " + text + linebreak
}

func printConcept(code string, description string) (result string) {
	result += "    Code:   " + code + linebreak
	result += "      Desc: " + description + linebreak
	return result
}

func errorCheck(err error) {
	if err != nil {
		panic(err)
	}
}
